product-service/src/database/connection.ts

```diff
@@ -0,0 +1,127 @@
+import {Pool, PoolClient} from 'pg';
+import * as dotenv from 'dotenv';
+import * as sqlQueries from './sql';
+
+// @ts-ignore
+dotenv.config( { silent: true });
+
+const {
+    PG_USER: user,
+    PG_HOST: host,
+    PG_DATABASE: database,
+    PG_PASSWORD: password,
+    PG_PORT: port
+} = process.env
+
+const databaseOptions = {
+    user,
+    host,
+    database,
+    password,
+    port: Number(port)
+}
+
+// cached pool
+let pool: Pool;
+
+export const PossibleQueries = {
+    SELECT_ALL_PRODUCTS: 0,
+    SELECT_PRODUCT_BY_ID: 1,
+    INSERT_PRODUCT: 2,
+    INSERT_STOCK: 3
+} as const;
+
+const queryOptionMap = {
+    [PossibleQueries.SELECT_ALL_PRODUCTS]: {
+        SQL: sqlQueries.getProducts
+    },
+    [PossibleQueries.SELECT_PRODUCT_BY_ID]: {
+        SQL: sqlQueries.getProductById
+    },
+    [PossibleQueries.INSERT_PRODUCT]: {
+        SQL: sqlQueries.insertProduct
+    },
+    [PossibleQueries.INSERT_STOCK]: {
+        SQL: sqlQueries.insertStock
+    }
+}
+
+type Keys = keyof typeof PossibleQueries;
+export type PossibleQueries = typeof PossibleQueries[Keys];
+
+const initPool = () => {
+    if (!pool) {
+        pool = new Pool(databaseOptions);
+    }
+
+    return pool;
+}
+
+const _requestSingle = (connection: PoolClient | undefined) => async <T> (query: PossibleQueries, params): Promise<T | null> => {
+
+    const result = await _requestArray(connection)<T>(query, params);
+
+    if(result.length === 0) {
+        return null;
+    }
+
+    return result[0];
+}
+
+const _requestArray = (connection: PoolClient | undefined) => async <T> (query: PossibleQueries, params): Promise<T[]> => {
+    const { SQL } = queryOptionMap[query];
+
+    if(connection) {
+        const result = await connection.query<T>(SQL, params);
+        return result.rows;
+    }
+    else {
+        const initializedPool = initPool();
+        const client = await initializedPool.connect();
+        try {
+            const result = await client.query<T>(SQL, params);
+            return result.rows;
+        }
+        catch (e) {
+            throw e;
+        }
+        finally {
+            // @ts-ignore
```

Comment:
What is the reason of `@ts-ignore` on every client.release?

Reply:
I used client.end method because release caused huge 10 seconds latency. 'end' method wasn't documented in typings and I got syntax errors from compilers. I figured out how to remove 10s latency and switched back to release method, but forgot to remove comments :)